### PR TITLE
feat: support EndpointSlice for the kubelet controller

### LINKED ISSUE #3862: Support the endpointslice role for Prometheus
**What is missing?**

The operator doesn't support the new `endpointslice` role available in Prometheus since v2.21.0 (https://github.com/prometheus/prometheus/pull/6838).

**Why do we need it?**

[Endpoint slices](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/) are required to scale services with many pods.

**Anything else we need to know?**:

After discussing on Slack with @brancz, the recommended approach would be that the operator uses seamlessly endpoint slices in lieu of endpoints when it detects that the Kube cluster supports them and Prometheus version is >= 2.21.0.

### PR DESCRIPTION
**Pull request by simonpasquier**
## Description

This change adds support for managing `EndpointSlice` objects for the kubelet service. The controller can manage either one of `Endpoints` and EndpointSlice` or both.
    
The migration path can be:
    1. Configure the operator to manage both objects in the kubelet controller.
    2. Verify that the generated `EndpointSlice` objects are correct.
    3. Configure the Prometheus object to use the `EndpointSlice` role instead of `Endpoints`.
    1. Configure the operator to manage only `Endpoints` objects in the kubelet controller.
    
The removal of the legacy `Endpoints` object is left to the user.

Closes #3862

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [X] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

File: Documentation/operator.md [hunk 1]
```diff
@@ -57,6 +57,10 @@ Usage of ./operator:
     	  PrometheusAgentDaemonSet: Enables the DaemonSet mode for PrometheusAgent (enabled: false)
   -key-file string
     	- NOT RECOMMENDED FOR PRODUCTION - Path to private TLS certificate file.
+  -kubelet-endpoints
```
Comment (nicolastakashi) [inline Documentation/operator.md:60]: Is it one or another right?
What I mean is we can't use both together?

File: example/rbac/prometheus-operator/prometheus-operator-deployment.yaml [hunk 2]
```diff
@@ -26,7 +26,7 @@ spec:
       containers:
       - args:
         - --kubelet-service=kube-system/kubelet
-        - --prometheus-config-reloader=quay.io/prometheus-operator/prometheus-config-reloader:v0.76.0
+        - --prometheus-config-reloader=quay.io/prometheus-operator/prometheus-config-reloader:v0.76.1
```
Comment (nicolastakashi) [inline example/rbac/prometheus-operator/prometheus-operator-deployment.yaml]: Interesting!
Should it be 0.76.2?

Comment (simonpasquier) [reply example/rbac/prometheus-operator/prometheus-operator-deployment.yaml]: It should be fixed now that conflicts are resolved.

**Comment by simonpasquier** [comment]
Since we have an upcoming release of the operator, I'd like to land this PR before. Given that nothing changes by default, it should be safe.
On Red Hat side, we will be turning on the endpoinslice flag in OpenShift and report back.

**Review by nicolastakashi — APPROVED** [review]

**Pull request merged**